lib/ansible/modules/windows/win_certificate_info.ps1

```diff
@@ -0,0 +1,145 @@
+#!powershell
+
+# Copyright: (c) 2019, Micah Hunsberger
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+function ConvertTo-Timestamp($start_date, $end_date)
+{
+    if ($start_date -and $end_date)
+    {
+        return (New-TimeSpan -Start $start_date -End $end_date).TotalSeconds
+    }
+}
+
+function Format-Date([DateTime]$date)
+{
+    return $date.ToUniversalTime().ToString('yyyy-MM-ddTHH:mm:ssK')
+}
+
+function Get-CertificateInfo ($cert)
+{
+    $epoch_date = Get-Date -Date "01/01/1970"
+
+    $cert_info = @{ }
+    $cert_info.friendly_name = $cert.FriendlyName
+    $cert_info.thumbprint = $cert.Thumbprint
+    $cert_info.subject = $cert.Subject
+    $cert_info.issuer = $cert.Issuer
+    $cert_info.valid_from = (ConvertTo-Timestamp -start_date $epoch_date -end_date $cert.NotBefore.ToUniversalTime())
+    $cert_info.valid_from_iso8601 = Format-Date -date $cert.NotBefore
+    $cert_info.valid_to = (ConvertTo-Timestamp -start_date $epoch_date -end_date $cert.NotAfter.ToUniversalTime())
+    $cert_info.valid_to_iso8601 = Format-Date -date $cert.NotAfter
+    $cert_info.serial_number = $cert.SerialNumber
+    $cert_info.archived = $cert.Archived
+    $cert_info.version = $cert.Version
+    $cert_info.has_private_key = $cert.HasPrivateKey
+    $cert_info.issued_by = $cert.GetNameInfo('SimpleName', $true)
+    $cert_info.issued_to = $cert.GetNameInfo('SimpleName', $false)
+    $cert_info.signature_algorithm = $cert.SignatureAlgorithm.FriendlyName
+    $cert_info.dns_names = @($cert_info.issued_to)
+    $cert_info.raw = [System.Convert]::ToBase64String($cert.GetRawCertData())
+    $cert_info.public_key = [System.Convert]::ToBase64String($cert.GetPublicKey())
+    $cert_info.extensions = @()
```

Comment:
Adding to an array can be quite an expensive operation, especially if the array elements are large, or there are lots of elements in the array. You are better of relying on PowerShell's piping mechanism that converts a collection to an array or build the result as an generic list and adding to it.

It's more PowerShell like to do the former and would look something like this

```
$cert_info.extensions = foreach($extension in $cert.Extensions) {
    $extension_info = @{
        critical = $extension.Critical
        field = $extension.Oid.FriendlyName
        value = $extension.Format($false)
    }
    .... # Do the rest of the work you want on the hashtable (if needed)

    # By outputting the hashtable you are adding it to the temp collection which is unpacked to an array
    # on the foreach loop. This means $cert_info.extensions is an array that contains each output you create
    $extension_info
}
```

Reply:
I did not know that `$result = foreach($elem in $array) { ... }` was valid, and thought you had to do the pipeline `$result = $array | Foreach-Object { ... }`. That does make sense since `$result = if($expr) { ... } else { ... }` and `$result = switch($item) { ... }` are both valid. Thanks!